server/service/src/sync/api/error.rs

```diff
@@ -1,58 +1,128 @@
 use super::*;
 use reqwest::{Response, StatusCode, Url};
-use serde::Deserialize;
+use serde::{
+    de::{value::StrDeserializer, IntoDeserializer},
+    Deserialize, Deserializer, Serialize, Serializer,
+};
 use thiserror::Error;
+use url::ParseError;
 
 #[derive(Error, Debug)]
-pub enum SyncApiError {
-    #[error("status: ({status}) error: {source}")]
-    MappedError {
-        source: SyncErrorV5,
-        status: StatusCode,
-    },
-    #[error("Error connecting to server: ({url})")]
-    ConnectionError { source: reqwest::Error, url: Url },
-    #[error("{0}")]
-    ResponseParsingError(ParsingResponseError),
-    #[error("{0}")]
-    Other(#[from] anyhow::Error),
+#[error("Sync api error, url: '{url}', route: '{route}'")]
+pub struct SyncApiError {
+    pub source: SyncApiErrorVariant,
+    pub(crate) url: Url,
+    pub(crate) route: String,
 }
 
-#[derive(Error, Debug, Deserialize)]
-pub enum SyncErrorV5 {
-    #[error("code: ({code}) message: ({message})")]
-    #[serde(rename = "error")]
+#[derive(Error, Debug)]
+pub enum SyncApiErrorVariant {
+    #[error("status: '{status}'")]
     ParsedError {
-        code: String,
-        message: String,
-        data: Option<String>,
+        status: StatusCode,
+        source: ParsedError,
+    },
+    #[error("status: '{status}' text: '{text}'")]
+    AsText { status: StatusCode, text: String },
+    #[error("Cannot parse error, status: '{status}'")]
+    ErrorParsingError {
+        status: StatusCode,
+        source: reqwest::Error,
     },
-    #[error("{0}")]
-    FullText(String),
+    #[error("Connection problem")]
+    ConnectionError(#[from] reqwest::Error),
+    #[error("Could not parse respose")]
+    ResponseParsingError(#[from] ParsingResponseError),
+    #[error("Could not parse url")]
+    FailToParseUrl(#[from] ParseError),
+    #[error("Unknown api error")]
+    Other(#[source] anyhow::Error),
 }
 
-impl SyncErrorV5 {
-    pub(crate) async fn from_response_and_status(
-        status: StatusCode,
-        response: Response,
-    ) -> SyncApiError {
-        let error = match to_json::<SyncErrorV5>(response).await {
-            Ok(source) => return SyncApiError::MappedError { source, status },
+#[derive(Error, Debug, Serialize, Deserialize)]
+#[error("code: '{code:?}' message: '{message}' data: '{data:?}")]
+pub struct ParsedError {
+    #[serde(serialize_with = "sync_error_code_v5_se")]
```

Comment:
In order to save any other sync api error codes in the `Other` variant of SyncErroCodeV5, used these two helpers